File: matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

double get_time()
{
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return tv.tv_sec + tv.tv_usec * 1e-6;
}

int main(int argc, char** argv)
{
  if (argc != 2) {
    printf("usage: %s N\n", argv[0]);
    return -1;
  }

  int n = atoi(argv[1]);
  double* a = (double*)malloc(n * n * sizeof(double)); // Matrix A
  double* b = (double*)malloc(n * n * sizeof(double)); // Matrix B
  double* c = (double*)malloc(n * n * sizeof(double)); // Matrix C

  // Initialize the matrices to some values.
  int i, j;
  for (i = 0; i < n; i++) {
    for (j = 0; j < n; j++) {
      a[i * n + j] = i * n + j; // A[i][j]
      b[i * n + j] = j * n + i; // B[i][j]
      c[i * n + j] = 0; // C[i][j]
    }
  }

  double begin = get_time();

  /**************************************/
  /* Write code to calculate C = A * B. */
  /**************************************/

  double end = get_time();
  printf("time: %.6lf sec\n", end - begin);

  // Print C for debugging. Comment out the print before measuring the execution time.
  double sum = 0;
  for (i = 0; i < n; i++) {
    for (j = 0; j < n; j++) {
      sum += c[i * n + j];
      // printf("c[%d][%d]=%lf\n", i, j, c[i * n + j]);
    }
  }
  // Print out the sum of all values in C.
  // This should be 450 for N=3, 3680 for N=4, and 18250 for N=5.
  printf("sum: %.6lf\n", sum);

  free(a);
  free(b);
  free(c);
  return 0;
}

File: matrix.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"time"
)

func newMatrix(n int) [][]int {
	m := make([][]int, n)
	for i := 0; i < n; i++ {
		m[i] = make([]int, n)
	}
	return m
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("usage: go run matrix.go N")
	}

	// Parse the N argument from the commandline.
	n, err := strconv.Atoi(os.Args[1])
	if err != nil || n <= 0 {
		log.Fatalf("need a positive in for N, not %v", os.Args[1])
	}

	// Initialize our matrices with some values.
	a := newMatrix(n)
	b := newMatrix(n)
	c := newMatrix(n)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			a[i][j] = i*n + j
			b[i][j] = j*n + i
			// c is already initialized to all zeros.
		}
	}

	begin := time.Now()

	//////////////////////////////////////////////
	// Write code here to calculate C = A * B
	// (without using numeric computation librarlies like gonum)
	//////////////////////////////////////////////

	end := time.Now()
	fmt.Println("time: ", end.Sub(begin))

	// Print contents of C for debugging.
	fmt.Println(c)

	total := 0
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			total += c[i][j]
		}
	}

	fmt.Println("sum: ", total)
	// This sum should be 450 for N=3, 3680 for N=4, and 18250 for N=5.
	knownSums := map[int]int{
		3: 450,
		4: 3680,
		5: 18250,
	}
	want := knownSums[n]
	if want != 0 && total != want {
		fmt.Printf("Wanted sum %v but got %v\n", want, total)
	}
}
